packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js

```diff
@@ -0,0 +1,158 @@
+      disablePublicKeys.forEach((id) => {
+        if (!identity.getPublicKeyById(id)) {
+          result.addError(
+            new InvalidIdentityPublicKeyIdError(id),
+          );
+        } else if (identity.getPublicKeyById(id).getReadOnly()) {
+          result.addError(
+            new IdentityPublicKeyIsReadOnlyError(id),
+          );
+        }
+      });
+
+      if (!result.isValid()) {
+        return result;
+      }
+
+      // Keys can only be disabled if another valid key is enabled in the same security level
+      disablePublicKeys.forEach(
+        (id) => identity.getPublicKeyById(id)
+          .setDisabledAt(stateTransition.getPublicKeysDisabledAt()),
+      );
+
+      const securityLevelsWithKeys = new Set();
+
+      SECURITY_LEVELS.forEach((securityLevel) => {
+        identity.getPublicKeys()
+          .filter((pk) => pk.getSecurityLevel() === securityLevel)
+          .forEach((pk) => {
+            if (!pk.getDisabledAt()) {
+              securityLevelsWithKeys.add(securityLevel);
+            }
+          });
+      });
+
+      SECURITY_LEVELS.forEach((securityLevel) => {
+        if (!securityLevelsWithKeys.has(securityLevel)) {
+          result.addError(new MissedSecurityLevelIdentityPublicKeyError(securityLevel));
+        }
+      });
+
+      if (!result.isValid()) {
+        return result;
+      }
+
+      // Calculate time window for timestamps
+      const {
+        time: {
+          seconds: lastBlockHeaderTimeSeconds,
+        },
+      } = await stateRepository.fetchLatestPlatformBlockHeader();
+
+      // Get last block header time in milliseconds
+      const lastBlockHeaderTime = lastBlockHeaderTimeSeconds * 1000;
+
+      // Define time window
+      const timeWindowStart = new Date(lastBlockHeaderTime);
+      timeWindowStart.setMinutes(
+        timeWindowStart.getMinutes() - BLOCK_TIME_WINDOW_MINUTES,
+      );
+
+      const timeWindowEnd = new Date(lastBlockHeaderTime);
+      timeWindowEnd.setMinutes(
+        timeWindowEnd.getMinutes() + BLOCK_TIME_WINDOW_MINUTES,
+      );
+
+      const disabledAtTime = stateTransition.getPublicKeysDisabledAt();
+
+      if (disabledAtTime < timeWindowStart || disabledAtTime > timeWindowEnd) {
+        result.addError(
+          new IdentityPublicKeyDisabledAtWindowViolationError(
+            disabledAtTime,
+            timeWindowStart,
+            timeWindowEnd,
+          ),
+        );
+      }
+
+      if (!result.isValid()) {
+        return result;
+      }
+    }
+
+    const addPublicKeys = stateTransition.getPublicKeysToAdd();
+    if (addPublicKeys) {
+      result.merge(
+        validatePublicKeysAreEnabled(addPublicKeys.map((pk) => pk.toObject())),
+      );
+
+      if (!result.isValid()) {
+        return result;
+      }
+
+      addPublicKeys.forEach((pk) => identity.getPublicKeys().push(pk));
+    }
+
+    result.merge(
+      validatePublicKeys(identity.getPublicKeys().map((pk) => pk.toObject())),
```

Comment:
I would recommend to validate disableAt here with JSON Schema instead of `validatePublicKeysAreEnabled`. You can just add an argument to `validatePublicKeys` to remove `disableAt` from schema.

Reply:
We already validated new keys, should we validate existing keys + new keys once again?